src/algebra/category/Module/change_of_rings.lean

```diff
@@ -0,0 +1,207 @@
+/-
+Copyright (c) 2022 Jujian Zhang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jujian Zhang
+-/
+import algebra.category.Module.basic
+import algebra.category.Algebra.basic
+import ring_theory.tensor_product
+
+/-!
+# Change Of Rings
+
+## Main definitions
+
+* `category_theory.Module.restrict_scalars`: given rings `R, S` and a ring homomorphism `R ⟶ S`,
+  then `restrict_scalars : Module S ⥤ Module R` is defined by `M ↦ M` where `M : S-module` is seen
+  as `R-module` by `r • m := f r • m` and `S`-linear map `l : M ⟶ M'` is `R`-linear as well.
+
+* `category_theory.Module.extend_scalars`: given **commutative** rings `R, S` and ring homomorphism
+  `f : R ⟶ S`, then `extend_scalars : Module R ⥤ Module S` is defined by `M ↦ S ⨂ M` where the
+  module structure is defined by `s • (s' ⊗ m) := (s * s') ⊗ m` and `R`-linear map `l : M ⟶ M'`
+  is sent to `S`-linear map `s ⊗ m ↦ s ⊗ l m : S ⨂ M ⟶ S ⨂ M'`.
+
+## List of notations
+Let `R, S` be rings and `f : R →+* S`
+* if `M` is an `R`-module, then notation `S ⨂[R, f] M` means the tensor product `S ⨂ M` where `S`
+  is considered as an `R`-module via restriction of scalars.
+* if `M` is an `R`-module, `s : S` and `m : M`, then `s ⊗ₜ[R, f]` is the pure tensor
+  `s ⊗ m : S ⊗[R, f] M`.
+-/
+
+
+namespace category_theory.Module
+
+universes v u₁ u₂
+
+namespace restrict_scalars
+
+variables {R : Type u₁} {S : Type u₂} [ring R] [ring S] (f : R →+* S)
+variable (M : Module.{v} S)
+
+/-- Any `S`-module M is also an `R`-module via a ring homomorphism `f : R ⟶ S` by defining
+    `r • m := f r • m` (`module.comp_hom`). This is called restriction of scalars. -/
+def obj' : Module R :=
+{ carrier := M,
+  is_module := module.comp_hom M f }
+
+/--
+Given an `S`-linear map `g : M → M'` between `S`-modules, `g` is also `R`-linear between `M` and
+`M'` by means of restriction of scalars.
+-/
+def map' {M M' : Module.{v} S} (g : M ⟶ M') :
+  obj' f M ⟶ obj' f M' :=
+{ map_smul' := λ r, g.map_smul (f r), ..g }
+
+/--
+If `R, S` are commutative rings and `f : R →+* S`, then any `S`-algebra is also an `R`-algebra
+-/
+def is_algebra {R : Type u₁} {S : Type u₂} [comm_ring R] [comm_ring S] (f : R →+* S)
+  (A : Type v) [comm_semiring A] [algebra S A] : algebra R A :=
+{ smul := (module.comp_hom A f).to_has_smul.smul,
+  to_fun := (algebra_map _ _).comp f,
+  map_one' := by simp,
+  map_mul' := λ _ _, by simp [map_mul],
+  map_zero' := by simp,
+  map_add' := λ _ _, by simp [map_add],
+  commutes' := λ _ _, by ring,
+  smul_def' := λ r a, algebra.smul_def _ _ }
```

Comment:
```suggestion
((algebra_map S A).comp f).to_algebra
```
I'm not sure if we need this as a separate definition though?

Reply:
This algebra instance/definition is only needed by `extend_scalars.map'`, but since it is so short, I will just inline it. Thank you